server/fleet/translator.go

```diff
@@ -0,0 +1,99 @@
+package fleet
+
+import (
+	"context"
+	"encoding/json"
+)
+
+const (
+	TranslatorTypeUserEmail = "User"
+	TranslatorTypeLabel     = "Label"
+	TranslatorTypeTeam      = "Team"
+	TranslatorTypeHost      = "Host"
+)
+
+type TranslatePayload struct {
+	Type    string          `json:"type"`
+	Payload json.RawMessage `json:"payload"`
+}
+
+type StringIdentifierToIDPayload struct {
+	Identifier string `json:"identifier"`
+	ID         uint   `json:"id"`
+}
+
+type IDGetter interface {
+	GetID() (uint, error)
+}
+
+type EmailToUserIDTranslator struct {
+	email string
+	ds    Datastore
+}
+
+func NewEmailToUserIDTranslator(ds Datastore, email string) *EmailToUserIDTranslator {
+	return &EmailToUserIDTranslator{email: email, ds: ds}
+}
+
+func (e *EmailToUserIDTranslator) GetID() (uint, error) {
+	user, err := e.ds.UserByEmail(e.email)
+	if err != nil {
+		return 0, err
+	}
+	return user.ID, nil
+}
```

Comment:
Sorry for all the churn on this, but it occurs to me seeing this version that it might be simplified to something like

```
type translateFunc func(ds Datastore, identifier string) (uint, error)

func translateEmailToUserID(ds Datastore, identifier string) (uint, error) {
	user, err := e.ds.UserByEmail(e.email)
	if err != nil {
		return 0, err
	}
	return user.ID, nil
}

// ... equivalent for other translations ...

func TranslateID(ds Datastore, translatorType string, identifier string) (uint, error) {
	var translator translateFunc
	switch translatorType {
	case TranslatorTypeUserEmail:
			translator = translateEmailToUserID
	// ... cases for other types ...
	default:
		return 0, errors.New("unknown translation type")
	}
	
	id, err := translator(ds, identifier)
	// ... handle error/return value ...
}
```

This seems potentially simpler to implement for new types, and simpler to call.

What do you think of a pattern like that?

Reply:
Yes! That looks way better. I'll give it a try.

Reply:
I also went ahead and removed the `Payload json.RawMessage` since we don't need it being generic now, and we might never need it to be generic. We can change it to that later if we ever have the need. So simplified further :D